mysql-test/suite/rpl/t/rpl_semi_sync_master_wait_for_slave_count.test

```diff
@@ -0,0 +1,111 @@
+--echo # MDEV-18983: `@@rpl_semi_sync_master_wait_for_slave_count` feature test
+# A gotcha: The FLUSH GLOBAL STATUS statments in this test are also replicated.
+
+--source include/have_debug.inc
+--source include/have_debug_sync.inc
+--source include/have_binlog_format_mixed.inc # format-agnostic
+# The variable's default is 1 to match the behavior before this feature came.
+# Therefore, this test uses two slaves to prove that the variable took effect.
+--let $rpl_topology= 1->2, 1->3
+--source include/rpl_init.inc
+
+
+--echo # Case 0: Semi-Sync commits only when both slaves ACKs
+# Simulate lag on one of the replicas
+--connection server_3
+SET @old_dbug= @@GLOBAL.debug_dbug;
+SET @@GLOBAL.debug_dbug= '+d,simulate_delay_semisync_slave_reply';
+
+# Send two transactions concurrently
+--connection server_1
+SET DEBUG_SYNC= 'semisync_got_reply SIGNAL wait WAIT_FOR continue';
+# Case 2.1 explicitly tests for a timeout, but this warning is also replicated
+--send CALL mtr.add_suppression('Timeout waiting for reply of binlog')
+  --connection default
+  SET DEBUG_SYNC= 'now WAIT_FOR wait';
+  SET DEBUG_SYNC= 'now SIGNAL continue';
+  --send CREATE TABLE t (a INT)
+
+    # Unblock lag for the first transcation on the first replica
+    --connection server_3
+    SET @@GLOBAL.debug_dbug= @old_dbug;
+    SET debug_sync= 'now SIGNAL io_thd_do_reply';
+    --connection server_1
+  --reap
+  --connection default
+--reap
+
+# Check status
+SHOW STATUS LIKE 'Rpl\_semi\_sync\_master\_%\_tx';
+SHOW STATUS LIKE 'Rpl\_semi\_sync\_master\_get\_ack';
+--connection server_2
+SHOW STATUS LIKE 'Rpl\_semi\_sync\_slave\_send\_ack';
+--connection server_3
+SHOW STATUS LIKE 'Rpl\_semi\_sync\_slave\_send\_ack';
+--connection server_1
+
+--echo # Case 1: Lowering a high requirement unblocks the waiting Semi-Sync
+SET @@GLOBAL.rpl_semi_sync_master_wait_for_slave_count= 3;
+SET DEBUG_SYNC= 'semisync_got_reply SIGNAL wait WAIT_FOR continue EXECUTE 2';
+--send FLUSH GLOBAL STATUS
+  --connection default
+  # Wait for 2 *complete* ACK processings
+  SET DEBUG_SYNC= 'now WAIT_FOR wait';
+  SET DEBUG_SYNC= 'now SIGNAL continue WAIT_FOR wait';
+  SET DEBUG_SYNC= 'now SIGNAL continue';
+  SHOW STATUS LIKE 'Rpl\_semi\_sync\_master\_%\_tx'; # expect 0
+  SET @@GLOBAL.rpl_semi_sync_master_wait_for_slave_count= 2;
+  --connection server_1
+--reap
+SHOW STATUS LIKE 'Rpl\_semi\_sync\_master\_%\_tx';
+
+
+--let $wait_no_slave= 0
+while ($wait_no_slave < 2)
+{
+  --echo # Cases 2.x Matrix:
+  --eval SET @@GLOBAL.rpl_semi_sync_master_wait_no_slave= $wait_no_slave
+
+  --echo # Case 2.1: falls back to Async when the requirement is set too high
```

Comment:
Should I `SET @@GLOBAL.rpl_semi_sync_master_timeout` to a short(er) wait?

Reply:
That's a good idea, it's always good to reduce test time, when you can.

Reply:
Except I’m not sure about how short to reduce it to, when our overloaded CI can time itself out just because it doesn’t have time to run a process.

Reply:
I think tests which rely on timing granularities of `2s` or lower are usually the ones that fail individually. Otherwise, when the CI overloads, usually the whole chunk of all tests all concurrently running on that builder will fail. So I'd bet anywhere in the range of `3s`-`5s` would be ok; then if we're wrong, we'll just adjust it :shrug:.

Reply:
> then if we're wrong, we'll just adjust it 🤷.

Many Hearthstone cards today get buffed in one patch but get reverted in the next immediate patch. 😁